napari/_app/injection/_processors.py

```diff
@@ -197,82 +149,16 @@ def _on_future_ready(f: Future):
-    type `napari.types.LayerData`, we return a function that can be called to
-    add a layer data tuple to the currently active viewer.
-    """
-    if type_ in _PROCESSORS:
-        return _PROCESSORS[type_]
-
-    if isinstance(type_, type):
-        for key, val in _PROCESSORS.items():
-            if isinstance(key, type) and issubclass(type_, key):
-                return val  # type: ignore [return-type]
-
-
-class set_processors:
-    """Set processor(s) for given type(s).
-
-    "Processors" are functions that can "do something" with an instance of the
-    type that they support.  For example, a processor that supports
-    `napari.types.ImageData` might take the data and add an Image layer to the
-    current viewer.
-
-    This is a class that behaves as a function or a context manager, that
-    allows one to set a processor function for a given type.
-
-    Parameters
-    ----------
-    mapping : Dict[Type[T], Callable[..., Optional[T]]]
-        a map of type -> processor function, where each value is a function
-        that is capable of retrieving an instance of the associated key/type.
-    clobber : bool, optional
-        Whether to override any existing processor function, by default False.
-
-    Raises
-    ------
-    ValueError
-        if clobber is `True` and one of the keys in `mapping` is already
-        registered.
-    """
-
-    def __init__(
-        self, mapping: Dict[Type[T], Callable[..., Optional[T]]], clobber=False
-    ):
-        self._before = {}
-        for k in mapping:
-            if k in _PROCESSORS and not clobber:
-                raise ValueError(
-                    f"Class {k} already has a processor and clobber is False"
-                )
-            self._before[k] = _PROCESSORS.get(k, _NULL)
-        _PROCESSORS.update(mapping)
-
-    def __enter__(self):
-        return None
-
-    def __exit__(self, *_):
-        for key, val in self._before.items():
-            if val is _NULL:
-                del _PROCESSORS[key]
-            else:
-                _PROCESSORS[key] = val
-
-
 # Add future and LayerData processors for each layer type.
-def _init_module():
-    for t in types._LayerData.__args__:
-        _PROCESSORS[t] = partial(_add_layer_data_to_viewer, return_type=t)
-
-        if sys.version_info >= (3, 9):
-            _PROCESSORS[Future[t]] = partial(
-                _add_future_data, return_type=t, _from_tuple=False
-            )
-
-
-_init_module()
+PROCESSORS: Dict[object, Callable] = {
+    types.LayerDataTuple: _add_layer_data_tuples_to_viewer,
+    List[types.LayerDataTuple]: _add_layer_data_tuples_to_viewer,
+    layers.Layer: _add_layer_to_viewer,
+}
+for t in types._LayerData.__args__:  # type: ignore [attr-defined]
+    PROCESSORS[t] = partial(_add_layer_data_to_viewer, return_type=t)
+
+    if sys.version_info >= (3, 9):
+        PROCESSORS[Future[t]] = partial(
+            _add_future_data, return_type=t, _from_tuple=False
+        )
```

Comment:
Did I good understand that it makes it harder to extend napari with its layer data type? At the last community meeting, we talk about such an option. maybe it could be dynamically calculated?

Reply:
i agree.  but this part is moved verbatim from another file already in `main` ... so would prefer to tackle that when we need to support arbitrary layers.  (i'd say we have a _lot_ to do to support layer extensions, well beyond this line)